### DISCUSSION TITLE
Track *non* computational dependencies & allow dummy nodes

### DISCUSSION THREAD

**Discussion by rgayler** [ropensci/targets#336] [Ideas]
@softloud and I are `targets` noobs and were discussing features we might find helpful. So there may be perfectly simple ways to do these things in `targets` that we just aren't aware of, in which case these things are "how to do it" questions, otherwise they are feature requests.

1. When using `targets`  to manage rendering of Rmarkdown/LaTeX/etc. template files, `targets` tracks the data inputs (via calls to `tar_load()`), the template file, and the rendered file. If the data inputs change, `targets` will flag the rendered file as out of date and requiring re-computation. However, the template file may contain manually entered text that also depends on the data inputs, e.g. subjective interpretation of plots generated from the input data.

  Would it be possible to register in `targets` a "manual" dependency of the template file on the data inputs so that when the data inputs change the template (and possibly downstream dependencies) is flagged as being manually out of date? That would alert the author to check whether the text content needs changing. There would need to be some manual method for the author to flag that the template is up to date.

2. We find the dependency graph from `tar_visnetwork()` to be very helpful for reasoning about the design of the computational pipeline. At the moment, the dependency graph reflects what we have actually implemented in our growing projects. We think it would be helpful if we could mock out nodes that we have yet to build.

  Would it be possible to create a target with its dependencies, but flagging the target as  dummy, so that `targets` doesn't try to execute it?

For both issues I expect we would get some extra columns in the manifest data frame, so we could write functions to track and support the use of these new targets status fields. It might even be useful to allow passing in some arbitrary key/value pairs when creating a target that would get carried into the manifest.

**Comment by wlandau** [comment]
## Literate programming

We are getting outside the scope of targets, but I can think of a hack that might still work. Sketch of a `_targets.R` file:

```r
library(targets)
library(tarchetypes)

run_and_write_indicator <- function() {
  on.exit(fs::file_create("indicator.log"))
  run_stuff_and_return_some_value()
}

prose_is_outdated <- file.exists("indicator.log")
cue <- tar_cue(mode = ifelse(prose_is_outdated, "always", "thorough"))

list(
  tar_target(dependency, run_and_write_indicator()),
  tar_render(report, "report.Rmd", cue = cue)
)
```

The idea is to write a special untracked indicator file when you run a dependency target tied to special prose. Whenever that indicator file exists in subsequent runs, the cue above will flag the report as always outdated and it will always rerun. That clue could nudge you to manually rewrite the affected prose and then delete the indicator file. But please be careful, this invalidation is only a clue. If you forget about this invalidation hack or forget how or why you set it up, that could lead to a lot of frustration, especially for your future self several months later.

## Sketching out pipelines

You can allow or exclude targets from running using the `names` argument of `tar_make()`, which supports selectors like `-starts_with("prefix_indicating_a_dummy_target")`. Alternatively, you can also decompose pipelines into separate lists of target objects

```r
# _targets.R file
library(targets)
pipeline1 <- list(tar_target(x, 1), tar_target(y, 2))
pipeline2 <- list(tar_target(z, 3), tar_target(w, 4))
list(pipeline1, pipeline2)
```

Above, if `pipeline1` is in production but `pipeline2` is a sketch, then you could comment out `pipeline2` for `tar_make()` and uncomment it for `tar_visnetwork()`.

**Reply by rgayler** [reply]
Thanks @wlandau .

We will try those out over the coming weeks and see how they feel.

**Reply by softloud** [reply]
> which supports selectors like -starts_with("prefix_indicating_a_dummy_target") 

:exploding_head: this is so, so awesome! Thank you.

**Reply by alexpghayes** [reply]
I think of this issue not so much as "literate programming" as it is "human in the loop" data analysis. This has come up for me in a slightly different context at https://github.com/ropensci/targets/discussions/419, where I can do my "human in the loop" actions via R. But in general I think tools for human in the loop steps would go a long way as elements of the `targets` ecosystem: the idea being that if you can't have a fully automated workflow, programmatic cues to add information are great. In my case, I'm clustering and I want to name the clusters myself, and then downstream targets need to use the cluster names.

**Reply by rgayler** [reply]
@alexpghayes yes, I think "human in the loop analysis" is a much more useful description. In my day-job the majority of the effort in an analytical project is spent on making data-based, but subjective, decisions on the following analytical steps (what sampling scheme to use, which predictive features to use, which model to use, etc.). When there are any upstream changes then the dependent subjective decisions should be reviewed to decide whether they are still appropriate.

The link to Rmarkdown that I focussed on in the OP is because the dominance of subjective decisions in the overall project forces a strong focus on documentation of the decisions made and the reasoning behind them, but the documentation is really dependent on the decisions and the reasoning that generates them.

I agree that having tools that identify when some manual action has been invalidated due to changes in upstream dependencies would help with my human in the loop data analysis. Your cluster labelling example should probably block downstream analysis until the manual labelling is completed. For my day-job work it would probably be more appropriate for all the downstream analyses to be re-run, but with the potentially invalidated nodes flagged for review (and a topological sort of the nodes so that I can review the topologically earliest results first).

**Reply by billdenney** [reply]
One of the things that I've done in the past to achieve similar results is to add a test within my report (it could be in another target, too).  Then, I do something like the following in my reporting.  I understand that it's not exactly what you're asking, but it does some of the catches for issues that may occur.  And, it also makes the reason for failure clearer.

```
    `r stopifnot(tar_read(parameter_x) > 0)` Parameter x goes up.
```
